transaction.go

```diff
@@ -49,198 +90,303 @@ type Transaction struct {
-				payloadsize = len(data[i*maxPayloadSize:])
-			}
-			pages[i].payload = make([]byte, payloadsize)
-			copy(pages[i].payload, data[i*maxPayloadSize:])
+	return buffer.Bytes(), nil
+}
+
+// unmarshalUpdates unmarshals the updates of a transaction
+func unmarshalUpdates(data []byte) ([]Update, error) {
+	buffer := bytes.NewBuffer(data)
+	updates := make([]Update, 0)
+
+	for {
+		update := Update{}
+
+		// Unmarshal name
+		var nameLength uint64
+		err1 := binary.Read(buffer, binary.LittleEndian, &nameLength)
+		if err1 == io.EOF {
+			// End of buffer reached
+			break
 		}
 
-		// Set the transaction number of the last page and increase the transactionCounter of the wal
-		pages[len(pages)-1].transactionNumber = t.wal.transactionCounter
-		t.wal.transactionCounter++
+		name := make([]byte, nameLength)
+		_, err2 := buffer.Read(name)
 
-		// Set the first and final page of the transaction
-		t.firstPage = &pages[0]
-		t.finalPage = &pages[len(pages)-1]
+		// Unmarshal version
+		var versionLength uint64
+		err3 := binary.Read(buffer, binary.LittleEndian, &versionLength)
 
-		// write the pages to disk and set the pageStatus
-		page := t.firstPage
-		for page != nil {
-			err := page.Write(t.wal.logFile)
-			if err != nil {
-				notifyChannel <- build.ExtendErr("Couldn't write the page to file", err)
-				return
-			}
-			page = page.nextPage
+		version := make([]byte, versionLength)
+		_, err4 := buffer.Read(version)
+
+		// Unmarshal instructions
+		var instructionsLength uint64
+		err5 := binary.Read(buffer, binary.LittleEndian, &instructionsLength)
+
+		instructions := make([]byte, instructionsLength)
+		_, err6 := buffer.Read(instructions)
+
+		// Check if any errors occured
+		if err1 != nil || err2 != nil || err3 != nil || err4 != nil || err5 != nil || err6 != nil {
+			return nil, errors.Compose(errors.New("Failed to unmarshal updates"), err1, err2, err3, err4, err5, err6)
 		}
 
-		// Commit transaction
-		notifyChannel <- t.commit()
-	}()
-	return notifyChannel
+		update.Name = string(name)
+		update.Version = string(version)
+		update.Instructions = instructions
+
+		updates = append(updates, update)
+	}
+
+	return updates, nil
 }
 
-// MarshalJSON marshales a transaction.
-func (t Transaction) checksum() ([crypto.HashSize]byte, error) {
+// threadedInitTransaction reserves pages of the wal, marshalls the
+// transactions's updates into a payload and splits the payload equally among
+// the pages. Once finished those pages are written to disk and the transaction
+// is committed.
+func threadedInitTransaction(t *Transaction) {
+	defer close(t.initComplete)
+
+	// Marshal all the updates to get their total length on disk
+	data, err := marshalUpdates(t.Updates)
+	if err != nil {
+		t.initComplete <- build.ExtendErr("could not marshal update", err)
+		return
+	}
+
+	// Find out how many pages are needed for the update
+	requiredPages := uint64(float64(len(data))/float64(maxPayloadSize+8) + 1)
```

Comment:
Not sure this is fully accurate. Might be better to calculate this out using a few lines of code:

```go
pagesNeeded := len(data)/maxPayloadSize
if len(data)%maxPayloadSize != 0 {
    pagesNeeded++
}
```
One thing that does confuse me here is the `+8` in the statement. I'm not sure where that's coming from, I think it should probably be removed.

Reply:
The `+8` is probably to account for the fact that you have to write the size of the payload before you write the payload.

Might need to have different constants.

Reply:
thinking a bit more, I think you just don't need the `+8` here